From 701a35232e9ed941311e91bdca1a39aa8d19c04a Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Tue, 25 Nov 2025 17:59:51 -0600
Subject: [PATCH 47/48] cmd: Implement mklink /j.

(cherry picked from commit 4ac592a323d736dd375fe9d90d368cb21483ad07)

CW-Bug-Id: #26492
---
 programs/cmd/builtins.c                  | 81 +++++++++++++++++++++++-
 programs/cmd/tests/test_builtins.bat     |  3 +
 programs/cmd/tests/test_builtins.bat.exp |  2 +
 programs/cmd/tests/test_builtins.cmd     |  3 +
 programs/cmd/tests/test_builtins.cmd.exp |  2 +
 5 files changed, 90 insertions(+), 1 deletion(-)

diff --git a/programs/cmd/builtins.c b/programs/cmd/builtins.c
index d7c5090d17f..c152c10b604 100644
--- a/programs/cmd/builtins.c
+++ b/programs/cmd/builtins.c
@@ -30,6 +30,10 @@
 
 #include "wcmd.h"
 #include <shellapi.h>
+#define WIN32_NO_STATUS
+#include "winternl.h"
+#include "winioctl.h"
+#include "ddk/ntifs.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(cmd);
@@ -3971,6 +3975,81 @@ RETURN_CODE WCMD_color(void)
   return errorlevel = return_code;
 }
 
+/* We cannot use SetVolumeMountPoint(), because that function forbids setting
+ * arbitrary directories as mount points, whereas mklink /j allows it. */
+BOOL create_mount_point(const WCHAR *link, const WCHAR *target) {
+    char buffer[MAXIMUM_REPARSE_DATA_BUFFER_SIZE];
+    REPARSE_DATA_BUFFER *data = (void *)buffer;
+    WCHAR full_link[MAX_PATH], *full_target;
+    UNICODE_STRING nt_link, nt_target;
+    OBJECT_ATTRIBUTES attr;
+    IO_STATUS_BLOCK io;
+    NTSTATUS status;
+    HANDLE file;
+    DWORD size;
+    BOOL ret;
+
+    TRACE( "link %s, target %s\n", debugstr_w(link), debugstr_w(target) );
+
+    if (!WCMD_get_fullpath(link, ARRAY_SIZE(full_link), full_link, NULL))
+        return FALSE;
+
+    if (!(size = GetFullPathNameW(target, 0, NULL, NULL)))
+        return FALSE;
+    full_target = malloc(size * sizeof(WCHAR));
+    GetFullPathNameW(target, size, full_target, NULL);
+
+    status = RtlDosPathNameToNtPathName_U_WithStatus(full_link, &nt_link, NULL, NULL);
+    if (status)
+    {
+        free(full_target);
+        return FALSE;
+    }
+
+    status = RtlDosPathNameToNtPathName_U_WithStatus(full_target, &nt_target, NULL, NULL);
+    if (status)
+    {
+        free(full_target);
+        RtlFreeUnicodeString(&nt_link);
+        return FALSE;
+    }
+
+    size = offsetof(REPARSE_DATA_BUFFER, MountPointReparseBuffer.PathBuffer);
+    size += nt_target.Length + sizeof(WCHAR) + (wcslen(full_target) + 1) * sizeof(WCHAR);
+    if (size > sizeof(buffer))
+    {
+        free(full_target);
+        RtlFreeUnicodeString(&nt_link);
+        RtlFreeUnicodeString(&nt_target);
+        return FALSE;
+    }
+
+    data->ReparseTag = IO_REPARSE_TAG_MOUNT_POINT;
+    data->ReparseDataLength = size - offsetof(REPARSE_DATA_BUFFER, MountPointReparseBuffer);
+    data->Reserved = 0;
+    data->MountPointReparseBuffer.SubstituteNameOffset = 0;
+    data->MountPointReparseBuffer.SubstituteNameLength = nt_target.Length;
+    data->MountPointReparseBuffer.PrintNameOffset = nt_target.Length + sizeof(WCHAR);
+    data->MountPointReparseBuffer.PrintNameLength = wcslen(full_target) * sizeof(WCHAR);
+    memcpy(data->MountPointReparseBuffer.PathBuffer,
+           nt_target.Buffer, nt_target.Length + sizeof(WCHAR));
+    memcpy(data->MountPointReparseBuffer.PathBuffer + (nt_target.Length / sizeof(WCHAR)) + 1,
+           full_target, (wcslen(full_target) + 1) * sizeof(WCHAR));
+    RtlFreeUnicodeString(&nt_target);
+    free(full_target);
+
+    InitializeObjectAttributes(&attr, &nt_link, OBJ_CASE_INSENSITIVE, 0, NULL);
+    status = NtCreateFile(&file, GENERIC_WRITE, &attr, &io, NULL, 0, 0, FILE_CREATE,
+                          FILE_OPEN_REPARSE_POINT | FILE_DIRECTORY_FILE, NULL, 0);
+    RtlFreeUnicodeString(&nt_link);
+    if (status)
+        return FALSE;
+
+    ret = DeviceIoControl(file, FSCTL_SET_REPARSE_POINT, data, size, NULL, 0, &size, NULL);
+    CloseHandle(file);
+    return ret;
+}
+
 /****************************************************************************
  * WCMD_mklink
  */
@@ -4021,7 +4100,7 @@ RETURN_CODE WCMD_mklink(WCHAR *args)
         else if(!junction)
             ret = CreateSymbolicLinkW(file1, file2, isdir);
         else
-            TRACE("Junction links currently not supported.\n");
+            ret = create_mount_point(file1, file2);
     }
 
     if (ret) return errorlevel = NO_ERROR;
diff --git a/programs/cmd/tests/test_builtins.bat b/programs/cmd/tests/test_builtins.bat
index 728a0e4eafc..ac716be57ae 100644
--- a/programs/cmd/tests/test_builtins.bat
+++ b/programs/cmd/tests/test_builtins.bat
@@ -229,6 +229,9 @@ call :setError 666 & (mklink &&echo SUCCESS !errorlevel!||echo FAILURE !errorlev
 call :setError 666 & (mklink /h foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
 call :setError 666 & (mklink /h foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
 call :setError 666 & (mklink /z foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
+call :setError 666 & (mklink /j foo foo >NUL &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
+call :setError 666 & (mklink /j foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
+rmdir foo
 echo bar > foo
 call :setError 666 & (mklink /h foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
 call :setError 666 & (mklink /h bar foo >NUL &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
diff --git a/programs/cmd/tests/test_builtins.bat.exp b/programs/cmd/tests/test_builtins.bat.exp
index a6b583f78c1..b26815a9cf2 100644
--- a/programs/cmd/tests/test_builtins.bat.exp
+++ b/programs/cmd/tests/test_builtins.bat.exp
@@ -142,6 +142,8 @@ FAILURE 1
 FAILURE 1
 FAILURE 1
 FAILURE 1
+SUCCESS 0
+FAILURE 1
 FAILURE 1
 SUCCESS 0
 FAILURE 1
diff --git a/programs/cmd/tests/test_builtins.cmd b/programs/cmd/tests/test_builtins.cmd
index 8552d51930a..a3c6a869703 100644
--- a/programs/cmd/tests/test_builtins.cmd
+++ b/programs/cmd/tests/test_builtins.cmd
@@ -691,6 +691,9 @@ call :setError 666 & (mklink &&echo SUCCESS !errorlevel!||echo FAILURE !errorlev
 call :setError 666 & (mklink /h foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
 call :setError 666 & (mklink /h foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
 call :setError 666 & (mklink /z foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
+call :setError 666 & (mklink /j foo foo >NUL &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
+call :setError 666 & (mklink /j foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
+rmdir foo
 echo bar > foo
 call :setError 666 & (mklink /h foo foo &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
 call :setError 666 & (mklink /h bar foo >NUL &&echo SUCCESS !errorlevel!||echo FAILURE !errorlevel!)
diff --git a/programs/cmd/tests/test_builtins.cmd.exp b/programs/cmd/tests/test_builtins.cmd.exp
index 292c24985e8..1b5c0c6209c 100644
--- a/programs/cmd/tests/test_builtins.cmd.exp
+++ b/programs/cmd/tests/test_builtins.cmd.exp
@@ -593,6 +593,8 @@ FAILURE 1
 FAILURE 1
 FAILURE 1
 FAILURE 1
+SUCCESS 0
+FAILURE 1
 FAILURE 1
 SUCCESS 0
 FAILURE 1
-- 
2.53.0

